File: src/com/cattailsw/nanidroid/install/NarDescriptorParser.java
package com.cattailsw.nanidroid.install;

import java.nio.ByteBuffer;
import java.nio.charset.CharacterCodingException;
import java.nio.charset.Charset;
import java.nio.charset.CharsetDecoder;
import java.nio.charset.CodingErrorAction;
import java.text.Normalizer;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** Pure parser for a defensively snapshotted NAR install descriptor. */
public final class NarDescriptorParser {
    private static final Charset ASCII = Charset.forName("US-ASCII");
    private static final Charset SHIFT_JIS = Charset.forName("Shift_JIS");
    private static final Charset UTF_8 = Charset.forName("UTF-8");
    private static final int MAX_DESCRIPTOR_BYTES = 64 * 1024;
    private static final int MAX_TARGET_BYTES = 255;
    private static final Pattern COMPOUND_INSTALL_KEY = Pattern.compile(
            "^(?:balloon|headline|plugin|calendar\\.skin|"
                    + "calendar\\.plugin)\\d*\\."
                    + "(directory|source\\.directory|refresh|"
                    + "refreshundeletemask)$");
    private static final Set<String> UNSUPPORTED_TYPES = unsupportedTypes();

    public NarDescriptorResult parse(
            byte[] descriptor,
            String forcedTargetId) {
        if (descriptor == null) {
            return NarDescriptorResult.failure(
                    NarInstallError.INVALID_METADATA, "null descriptor");
        }
        if (descriptor.length > MAX_DESCRIPTOR_BYTES) {
            return NarDescriptorResult.failure(
                    NarInstallError.INSTALL_DESCRIPTOR_LIMIT,
                    "descriptor exceeds 64 KiB");
        }
        byte[] snapshot = snapshot(descriptor);
        try {
            return NarDescriptorResult.success(
                    parseSnapshot(snapshot, forcedTargetId));
        } catch (Rejected rejected) {
            return NarDescriptorResult.failure(
                    rejected.error, rejected.getMessage());
        }
    }

    private static NarInstallDescriptor parseSnapshot(
            byte[] bytes,
            String forcedTargetId) throws Rejected {
        Encoding encoding = selectEncoding(bytes);
        String text = decode(bytes, encoding.offset, encoding.charset);
        Map<String, String> metadata = parseLines(text);
        metadata.put("charset", encoding.charset.name());
        rejectCompoundInstall(metadata);

        if (!metadata.containsKey("type")) {
            reject(NarInstallError.MISSING_TYPE, "type is required");
        }
        String type = metadata.get("type");
        if (type.length() == 0) {
            reject(NarInstallError.INVALID_TYPE, "blank type");
        }
        String normalizedType = collisionKey(type);
        if (!"ghost".equals(normalizedType)) {
            reject(
                    UNSUPPORTED_TYPES.contains(normalizedType)
                            ? NarInstallError.UNSUPPORTED_TYPE
                            : NarInstallError.INVALID_TYPE,
                    normalizedType);
        }
        if (!metadata.containsKey("name")
                || metadata.get("name").length() == 0
                || !metadata.containsKey("directory")
                || metadata.get("directory").length() == 0) {
            reject(
                    NarInstallError.MISSING_METADATA,
                    "name and directory are required");
        }

        String descriptorDirectory = normalizeTarget(
                metadata.get("directory"));
        if (descriptorDirectory == null) {
            reject(NarInstallError.INVALID_TARGET_ID, "unsafe directory");
        }
        String targetId = forcedTargetId == null
                ? descriptorDirectory
                : normalizeTarget(forcedTargetId);
        if (targetId == null) {
            reject(NarInstallError.INVALID_TARGET_ID, "unsafe forced id");
        }
        if ("1".equals(metadata.get("refresh"))) {
            reject(NarInstallError.UNSUPPORTED_REFRESH, "unsupported");
        }

        metadata.put("type", "ghost");
        metadata.put("directory", descriptorDirectory);
        return new NarInstallDescriptor(
                "ghost",
                metadata.get("name"),
                descriptorDirectory,
                targetId,
                metadata.get("accept"),
                metadata);
    }

    static byte[] snapshot(byte[] descriptor) {
        return descriptor.clone();
    }

    private static void rejectCompoundInstall(Map<String, String> metadata)
            throws Rejected {
        boolean compoundInstall = false;
        for (Map.Entry<String, String> entry : metadata.entrySet()) {
            Matcher matcher = COMPOUND_INSTALL_KEY.matcher(entry.getKey());
            if (!matcher.matches()) {
                continue;
            }
            compoundInstall = true;
            if ("refresh".equals(matcher.group(1))
                    && "1".equals(entry.getValue())) {
                reject(
                        NarInstallError.UNSUPPORTED_REFRESH,
                        entry.getKey());
            }
        }
        if (compoundInstall) {
            reject(
                    NarInstallError.UNSUPPORTED_COMPOUND_INSTALL,
                    "compound install directive");
        }
    }

    private static Encoding selectEncoding(byte[] bytes) throws Rejected {
        if (bytes.length >= 3
                && (bytes[0] & 0xff) == 0xef
                && (bytes[1] & 0xff) == 0xbb
                && (bytes[2] & 0xff) == 0xbf) {
            return new Encoding(UTF_8, 3);
        }
        int end = 0;
        while (end < bytes.length
                && bytes[end] != '\n'
                && bytes[end] != '\r') {
            end++;
        }
        String firstLine = new String(bytes, 0, end, ASCII);
        int comma = firstLine.indexOf(',');
        if (comma > 0
                && "charset".equals(collisionKey(
                        firstLine.substring(0, comma).trim()))) {
            String name = firstLine.substring(comma + 1).trim();
            try {
                return new Encoding(Charset.forName(name), 0);
            } catch (RuntimeException error) {
                reject(
                        NarInstallError.UNSUPPORTED_DESCRIPTOR_CHARSET,
                        name);
            }
        }
        return new Encoding(SHIFT_JIS, 0);
    }

    private static String decode(
            byte[] bytes,
            int offset,
            Charset charset) throws Rejected {
        try {
            CharsetDecoder decoder = charset.newDecoder()
                    .onMalformedInput(CodingErrorAction.REPORT)
                    .onUnmappableCharacter(CodingErrorAction.REPORT);
            return decoder.decode(
                    ByteBuffer.wrap(bytes, offset, bytes.length - offset))
                    .toString();
        } catch (CharacterCodingException error) {
            reject(
                    NarInstallError.INVALID_DESCRIPTOR_ENCODING,
                    charset.name());
            return null;
        }
    }

    private static Map<String, String> parseLines(String text)
            throws Rejected {
        Map<String, String> metadata =
                new LinkedHashMap<String, String>();
        String[] lines = text.split("\\r?\\n", -1);
        for (int index = 0; index < lines.length; index++) {
            String line = lines[index];
            if (line.trim().length() == 0) {
                continue;
            }
            int comma = line.indexOf(',');
            if (comma <= 0) {
                reject(NarInstallError.INVALID_METADATA, "malformed line");
            }
            String key = collisionKey(line.substring(0, comma).trim());
            String value = Normalizer.normalize(
                    line.substring(comma + 1).trim(),
                    Normalizer.Form.NFC);
            if (key.length() == 0
                    || containsControl(key)
                    || containsControl(value)
                    || ("charset".equals(key) && index != 0)
                    || metadata.containsKey(key)) {
                reject(
                        NarInstallError.INVALID_METADATA,
                        "invalid or duplicate metadata");
            }
            metadata.put(key, value);
        }
        return metadata;
    }

    private static String normalizeTarget(String value) {
        if (value == null
                || value.length() == 0
                || value.length() > MAX_TARGET_BYTES
                || !validUnicode(value)
                || hasBoundaryWhitespace(value)
                || value.indexOf('/') >= 0
                || value.indexOf('\\') >= 0
                || value.indexOf(':') >= 0
                || containsControl(value)
                || ".".equals(value)
                || "..".equals(value)) {
            return null;
        }
        String normalized = Normalizer.normalize(value, Normalizer.Form.NFC);
        return normalized.getBytes(UTF_8).length <= MAX_TARGET_BYTES
                ? normalized
                : null;
    }

    private static boolean hasBoundaryWhitespace(String value) {
        int first = value.codePointAt(0);
        int last = value.codePointBefore(value.length());
        return Character.isWhitespace(first)
                || Character.isSpaceChar(first)
                || Character.isWhitespace(last)
                || Character.isSpaceChar(last);
    }

    private static String collisionKey(String value) {
        String nfc = Normalizer.normalize(value, Normalizer.Form.NFC);
        return Normalizer.normalize(
                nfc.toLowerCase(Locale.US),
                Normalizer.Form.NFC);
    }

    private static boolean containsControl(String value) {
        for (int index = 0; index < value.length(); index++) {
            if (Character.isISOControl(value.charAt(index))) {
                return true;
            }
        }
        return false;
    }

    private static boolean validUnicode(String value) {
        for (int index = 0; index < value.length(); index++) {
            char current = value.charAt(index);
            if (Character.isHighSurrogate(current)) {
                if (++index >= value.length()
                        || !Character.isLowSurrogate(value.charAt(index))) {
                    return false;
                }
            } else if (Character.isLowSurrogate(current)) {
                return false;
            }
        }
        return true;
    }

    private static Set<String> unsupportedTypes() {
        Set<String> types = new HashSet<String>();
        types.addAll(Arrays.asList(
                "shell",
                "supplement",
                "balloon",
                "plugin",
                "headline",
                "language",
                "calendar skin",
                "calendar plugin",
                "calendar",
                "package"));
        return Collections.unmodifiableSet(types);
    }

    private static void reject(NarInstallError error, String detail)
            throws Rejected {
        throw new Rejected(error, detail);
    }

    private static final class Encoding {
        private final Charset charset;
        private final int offset;
        private Encoding(Charset charset, int offset) {
            this.charset = charset;
            this.offset = offset;
        }
    }

    private static final class Rejected extends Exception {
        private final NarInstallError error;
        private Rejected(NarInstallError error, String detail) {
            super(detail);
            this.error = error;
        }
    }
}

File: src/com/cattailsw/nanidroid/install/NarDescriptorParser.kt
package com.cattailsw.nanidroid.install

import java.nio.ByteBuffer
import java.nio.charset.CharacterCodingException
import java.nio.charset.Charset
import java.nio.charset.CodingErrorAction
import java.text.Normalizer
import java.util.Locale

/** Pure parser for a defensively snapshotted NAR install descriptor. */
class NarDescriptorParser {
    fun parse(descriptor: ByteArray?, forcedTargetId: String?): NarDescriptorResult {
        if (descriptor == null) return NarDescriptorResult.failure(NarInstallError.INVALID_METADATA, "null descriptor")
        if (descriptor.size > MAX_DESCRIPTOR_BYTES) return NarDescriptorResult.failure(NarInstallError.INSTALL_DESCRIPTOR_LIMIT, "descriptor exceeds 64 KiB")
        return try {
            NarDescriptorResult.success(parseSnapshot(descriptor.copyOf(), forcedTargetId))
        } catch (rejected: Rejected) {
            NarDescriptorResult.failure(rejected.error, rejected.message!!)
        }
    }

    @Throws(Rejected::class)
    private fun parseSnapshot(bytes: ByteArray, forcedTargetId: String?): NarInstallDescriptor {
        val encoding = selectEncoding(bytes)
        val metadata = parseLines(decode(bytes, encoding.offset, encoding.charset))
        metadata["charset"] = encoding.charset.name()
        val type = metadata["type"] ?: reject(NarInstallError.MISSING_TYPE, "type is required")
        if (type.isEmpty()) reject(NarInstallError.INVALID_TYPE, "blank type")
        val normalizedType = collisionKey(type)
        if (normalizedType != "ghost") reject(if (normalizedType in UNSUPPORTED_TYPES) NarInstallError.UNSUPPORTED_TYPE else NarInstallError.INVALID_TYPE, normalizedType)
        if (metadata["name"].isNullOrEmpty() || metadata["directory"].isNullOrEmpty()) reject(NarInstallError.MISSING_METADATA, "name and directory are required")
        val descriptorDirectory = normalizeTarget(metadata["directory"]) ?: reject(NarInstallError.INVALID_TARGET_ID, "unsafe directory")
        val targetId = if (forcedTargetId == null) descriptorDirectory else normalizeTarget(forcedTargetId) ?: reject(NarInstallError.INVALID_TARGET_ID, "unsafe forced id")
        // Fresh installs extract a complete archive into a new target. Preserve
        // refresh and companion metadata for runtime compatibility; update code
        // continues to apply its own refresh restrictions.
        metadata["type"] = "ghost"
        metadata["directory"] = descriptorDirectory
        return NarInstallDescriptor("ghost", metadata["name"]!!, descriptorDirectory, targetId, metadata["accept"], metadata)
    }

    @Throws(Rejected::class)
    private fun selectEncoding(bytes: ByteArray): Encoding {
        if (bytes.size >= 3 && (bytes[0].toInt() and 0xff) == 0xef && (bytes[1].toInt() and 0xff) == 0xbb && (bytes[2].toInt() and 0xff) == 0xbf) return Encoding(UTF_8, 3)
        var end = 0
        while (end < bytes.size && bytes[end] != '\n'.code.toByte() && bytes[end] != '\r'.code.toByte()) end++
        val firstLine = String(bytes, 0, end, ASCII)
        val comma = firstLine.indexOf(',')
        if (comma > 0 && collisionKey(javaTrim(firstLine.substring(0, comma))) == "charset") {
            val name = javaTrim(firstLine.substring(comma + 1))
            return try { Encoding(Charset.forName(name), 0) } catch (_: RuntimeException) { reject(NarInstallError.UNSUPPORTED_DESCRIPTOR_CHARSET, name) }
        }
        return Encoding(SHIFT_JIS, 0)
    }

    @Throws(Rejected::class)
    private fun decode(bytes: ByteArray, offset: Int, charset: Charset): String = try {
        charset.newDecoder().onMalformedInput(CodingErrorAction.REPORT).onUnmappableCharacter(CodingErrorAction.REPORT).decode(ByteBuffer.wrap(bytes, offset, bytes.size - offset)).toString()
    } catch (_: CharacterCodingException) { reject(NarInstallError.INVALID_DESCRIPTOR_ENCODING, charset.name()) }

    @Throws(Rejected::class)
    private fun parseLines(text: String): LinkedHashMap<String, String> {
        val metadata = linkedMapOf<String, String>()
        // Java used a negative split limit, which retains trailing empty
        // fields. Kotlin requires a non-negative limit; Int.MAX_VALUE is the
        // equivalent unbounded, trailing-field-preserving limit.
        text.split(Regex("\\r?\\n"), Int.MAX_VALUE).forEach { line ->
            if (javaTrim(line).isEmpty()) return@forEach
            val comma = line.indexOf(',')
            if (comma <= 0) reject(NarInstallError.INVALID_METADATA, "malformed line")
            val key = collisionKey(javaTrim(line.substring(0, comma)))
            val value = Normalizer.normalize(javaTrim(line.substring(comma + 1)), Normalizer.Form.NFC)
            if (key.isEmpty() || containsControl(key) || containsControl(value) || (key == "charset" && metadata.isNotEmpty()) || metadata.containsKey(key)) reject(NarInstallError.INVALID_METADATA, "invalid or duplicate metadata")
            metadata[key] = value
        }
        return LinkedHashMap(metadata)
    }


    private fun normalizeTarget(value: String?): String? {
        if (value == null || value.isEmpty() || value.length > MAX_TARGET_BYTES || !validUnicode(value) || hasBoundaryWhitespace(value) || '/' in value || '\\' in value || ':' in value || containsControl(value) || value == "." || value == "..") return null
        val normalized = Normalizer.normalize(value, Normalizer.Form.NFC)
        return if (normalized.toByteArray(UTF_8).size <= MAX_TARGET_BYTES) normalized else null
    }
    /** Exact Java String.trim() behavior: only UTF-16 code units <= U+0020. */
    private fun javaTrim(value: String): String = value.trim { it <= '\u0020' }
    private fun hasBoundaryWhitespace(value: String): Boolean { val first = value.codePointAt(0); val last = value.codePointBefore(value.length); return Character.isWhitespace(first) || Character.isSpaceChar(first) || Character.isWhitespace(last) || Character.isSpaceChar(last) }
    private fun collisionKey(value: String): String = Normalizer.normalize(Normalizer.normalize(value, Normalizer.Form.NFC).lowercase(Locale.US), Normalizer.Form.NFC)
    private fun containsControl(value: String): Boolean = value.any { Character.isISOControl(it) }
    private fun validUnicode(value: String): Boolean { var index = 0; while (index < value.length) { val current = value[index]; if (Character.isHighSurrogate(current)) { if (++index >= value.length || !Character.isLowSurrogate(value[index])) return false } else if (Character.isLowSurrogate(current)) return false; index++ }; return true }
    @Throws(Rejected::class) private fun reject(error: NarInstallError, detail: String): Nothing = throw Rejected(error, detail)
    private class Encoding(val charset: Charset, val offset: Int)
    private class Rejected(val error: NarInstallError, detail: String) : Exception(detail)
    private companion object {
        val ASCII: Charset = Charset.forName("US-ASCII"); val SHIFT_JIS: Charset = Charset.forName("Shift_JIS"); val UTF_8: Charset = Charset.forName("UTF-8")
        const val MAX_DESCRIPTOR_BYTES = 64 * 1024; const val MAX_TARGET_BYTES = 255
        val UNSUPPORTED_TYPES = setOf("shell", "supplement", "balloon", "plugin", "headline", "language", "calendar skin", "calendar plugin", "calendar", "package")
    }
}
